apps/src/applab/GoogleChart.js

```diff
@@ -0,0 +1,463 @@
+ newcap: true,
+ nonew: true,
+ shadow: false,
+ unused: true,
+ eqeqeq: true,
+
+ maxlen: 90,
+ maxparams: 6,
+ maxstatements: 200
+ */
+'use strict';
+/* global google */
+
+var Promise = require('es6-promise').Promise;
+require('../utils');
+
+/**
+ * Abstract base type for Google Charts API charts.
+ *
+ * @example
+ *   var chart = new PieChart(targetDiv);
+ *   Promise.resolve(function () {
+ *     return chart.loadDependencies();
+ *   }).then(function () {
+ *     return chart.loadData(data[, columns]);
+ *   }).then(function () {
+ *     return chart.drawChart([options]);
+ *   });
+ *
+ * May optionally omit the loadDependencies() step; if so, it will automatically
+ * occur as needed prior to another asynchronous step.  It's exposed here to
+ * facilitate asynchronous loading.
+ *
+ * @param {Element} targetDiv
+ * @constructor
+ */
+var GoogleChart = module.exports = function (targetDiv) {
+  // Define this late so we can overwrite it in tests.
+  if (!GoogleChart.lib) {
+    GoogleChart.lib = google;
+  }
+
+  /** @private {Element} */
+  this.targetDiv_ = targetDiv;
+
+  /** @private {google.visualization.DataTable} */
+  this.dataTable_ = null;
+
+  /**
+   * List of all warnings logged while performing operations with this chart
+   * instance.
+   * @type {Error[]}
+   */
+  this.warnings = [];
+};
+
+/**
+ * Loads the required libraries for this particular chart type.
+ * Safe to call multiple times - Google's loader caches dependencies.
+ * @returns {Promise} that resolves when dependencies have been loaded.
+ */
+GoogleChart.prototype.loadDependencies = function () {
+  return new Promise(function (resolve, reject) {
+    try {
+      GoogleChart.lib.load('visualization', '1', {
+        packages: this.getDependencies(),
+        callback: resolve
+      });
+    } catch (e) {
+      // We catch and return a different error so that we don't surface Google
+      // API errors to students.
+      reject(new Error('Unable to load Charts API.  Please try again later.'));
+    }
+  }.bind(this));
+};
+
+/**
+ * Validates and prepares the provided data to be displayed by the chart.
+ *
+ * Must be called prior to drawChart().  If loadDependencies() has not been
+ * called before this method, dependencies will be loaded now.
+ *
+ * @param {Object[]} rawData - data to display in chart, formatted as an array
+ *        of objects where each object represents a row, and the object keys
+ *        are column names.
+ * @param {string[]} [columnList] - Ordered list of column names to use as source
+ *        data for the chart.  Column names must match keys in rawData.  If
+ *        omitted, column names will be inferred from rawData.
+ * @returns {Promise} that resolves when the chart data has been loaded and
+ *          the chart is ready for drawChart().
+ */
+GoogleChart.prototype.loadData = function (rawData, columnList) {
+  // If columnList is not provided, infer it from raw data
+  if (columnList === undefined) {
```

Comment:
What is the desired behavior when columnList is empty? It seems like we might want to print a warning rather than displaying an empty chart (assuming that's what happens when you pass Google Charts an empty list of columns).